lib/ex_doc/cli.ex

```diff
@@ -35,14 +35,25 @@ defmodule ExDoc.CLI do
           quiet: :boolean,
           source_ref: :string,
           source_url: :string,
-          version: :boolean
+          version: :boolean,
+          warnings_as_errors: :boolean
         ]
       )
 
-    if List.keymember?(opts, :version, 0) do
-      IO.puts("ExDoc v#{ExDoc.version()}")
-    else
-      generate(args, opts, generator)
+    cond do
+      List.keymember?(opts, :version, 0) ->
+        IO.puts("ExDoc v#{ExDoc.version()}")
+
+      opts[:warnings_as_errors] == true and ExDoc.Utils.warned?() ->
+        IO.puts(
+          :stderr,
+          "Doc generation failed due to warnings while using the --warnings-as-errors option"
+        )
+
+        exit({:shutdown, 1})
+
+      true ->
+        generate(args, opts, generator)
```

Comment:
we are checking for warnings without generating the docs. I believe this should rather be something like this:

```elixir
if List.keymember?(opts, :version, 0) do
  IO.puts("ExDoc v#{ExDoc.version()}")
else
  generate(args, opts, generator)

  if opts[:warnings_as_errors] == true and ExDoc.Utils.warned?() do
    # ...
  end
end
```

we need to improve cli_test.exs because it's not catching this bug.